File: Python/other/Russian_Peasant_Algorithm.py
# Python Program to multiply two numbers using Russian Peasant method 

# Function to multiply two numbers using Russian Peasant method 
def russianPeasant(number1, number2): 

	result = 0 

	# While second number doesn't 
	# become 1 
	while (number2 > 0): 
	
		# If second number becomes odd, add the first number to result 
		if (number2 & 1): 
			result = result + number1 

		# Double the first number and halve the second number 
		number1 = number1 << 1 		 #bitwise left operator
		number2 = number2 >> 1 		 #bitwise right operator
	
	return print("Result is",result )

# Driver program to test 
# above function 
number1=int(input("Enter 1st Number:"))
number2=int(input("Enter 2nd Number:"))
russianPeasant(number1,number2) 

'''
Sample I/O:
	Input:
	Enter 1st Number:5
	Enter 2nd Number:5

	Output:
	Result is 25

Time complexity: Θ(1)
Space complexity: Θ(1)
'''

File: C/other/Russian_Peasant_Algorithm.c
/*
In this algorithm we keep halving one number and doubling the 2nd one.
While halving, if we get an odd number, we add the 2nd doubled number to the result.
This goes on until the halving makes the first reach 0.
It is also known as the Ancient Egyptian Multiplication, used by
scribes to do multiplication without using the multiplication table.
*/
#include <stdio.h>

int Russian_Peasant_Algo(int, int);
int main()
{
	int a, b, ans;
	printf("Enter the first number: ");
	scanf("%d", &a);
	printf("\nEnter the second number: ");
	scanf("%d", &b);
	ans = Russian_Peasant_Algo(a, b);
	printf("\nThe product of the given number is %d\n", ans);
	return 0;
}
// Implementing the Russian-Peasant Algorithm
int Russian_Peasant_Algo(int a, int b)
{
	int ans = 0;
	while (b)
	{
		if (b % 2)
		{
			ans += a;
		}
		a *= 2;
		b /= 2;
	}
	return ans;
}

/*
Time Complexity: O(log2(max(no of digits in a, no of digits in b)))

Sample Output
Enter the first number: 4
Enter the second number: 5
The product of the given numbers is 20
*/

